### PR TITLE
Add analytic dipole derivatives for RHF

### PR DESCRIPTION
**Pull request by andysim**
## Description
Adds analytic dipole derivatives for RHF, allowing IR intensities to be computed; @loriab, @dsirianni and crew are working on the finite difference version for cases like DF-MP2, where finite differences of gradients are used for frequencies.

## Todos
Notable points (developer or user-interest) that this PR has or will accomplish.
- [x] Adds dipole derivative code, and a test to verify that they match the CCQC programming project values

## Questions
- [ ] Does anybody object to wavefunction holding these?
- [ ] Does anybody object to the various `set_` methods of wavefunction taking SharedMatrix by value, not by reference?

## Checklist
- [x] Tests added for any new features
- [x] [All or relevant fraction of full tests run](http://psicode.org/psi4manual/master/build_planning.html#how-to-run-a-subset-of-tests)

## Status
- [x] Ready for review
- [x] Ready for merge

### CODE REVIEW

File: psi4/src/export_wavefunction.cc [hunk 1]
```diff
@@ -129,6 +129,10 @@ void export_wavefunction(py::module& m) {
         .def("get_basisset", &Wavefunction::get_basisset, "Returns the requested auxiliary basis.")
         .def("set_basisset", &Wavefunction::set_basisset, "Sets the requested auxiliary basis.")
         .def("energy", &Wavefunction::reference_energy, "Returns the Wavefunctions energy.")
+        .def("dipole_gradient", &Wavefunction::dipole_gradient,
+             "Returns the Wavefunctions dipole derivative w.r.t. nuclear perturbations (a.u.) as a 3N x 3 matrix.")
+        .def("set_dipole_gradient", &Wavefunction::set_dipole_gradient,
```
Comment (dgasmith) [inline psi4/src/export_wavefunction.cc]: Can we hold these in the `get/set_array` generics?

File: psi4/src/psi4/libmints/mintshelper.cc [hunk 2]
```diff
@@ -1995,6 +1996,171 @@ SharedMatrix MintsHelper::perturb_grad(SharedMatrix D, double xlambda, double yl
...
+    double **Dp = D->pointer();
+
+    int natom = molecule_->natom();
+    auto ret = std::make_shared<Matrix>("Dipole dervatives (pert*component, i.e. 3Nx3)", 3 * natom, 3);
+    double **Pp = ret->pointer();
+
+    std::shared_ptr<OneBodyAOInt> Dint(integral_->ao_dipole(1));
+    const double *buffer = Dint->buffer();
+
+    for (int P = 0; P < basisset_->nshell(); P++) {
```
Comment (dgasmith) [inline psi4/src/psi4/libmints/mintshelper.cc]: I think we copy/paste this code a bit. Is it possible to make a general function that takes a OneBodyAOInt derivative's to produce this quantity?

File: psi4/src/export_wavefunction.cc (hunk 1 above)
Comment (loriab) [reply psi4/src/export_wavefunction.cc]: if it helps the decision making, I think the named arrays (energy, gradient, hessian, dipole_gradient?) are (1) the things we set `CURRENT` for and (2) the things that, on the wfn, should always be a synched level of theory. A related question is whether after qcvar reform completed, E/G/H on wfn goes away and qcvars on wfn w/ a complement of `CURRENT`s take over or whether E/G/H on wfn continues and `CURRENT`s dropped altogether.

File: psi4/src/psi4/libmints/mintshelper.cc [hunk 3]
```diff
@@ -1812,188 +1812,13 @@ SharedMatrix MintsHelper::perturb_grad(SharedMatrix D) {
...
-                }
-            }
-        }
-    }
-    return ret;
+    int natoms = basisset_->molecule()->natom();
+    auto perturbation_gradient = std::make_shared<Matrix>("Perturbation Gradient", natoms, 3);
+    auto dipole_gradients = dipole_grad(D);
+    double lambdas[3] = {xlambda, ylambda, zlambda};
+    C_DGEMM('n', 't', 3 * natoms, 1, 3, 1.0, dipole_gradients->pointer()[0], 3, &lambdas[0], 3, 0.0,
```
Comment (dgasmith) [inline psi4/src/psi4/libmints/mintshelper.cc]: Well, that's easier.

Comment (andysim) [reply psi4/src/psi4/libmints/mintshelper.cc]: Sure is.  The pert_grad routine computed this contraction within the loops, so wasn't suitable for the dipole gradients.  By computing the uncontracted dipole gradients and GEMMing them, we avoid repetition as you suggested - good call.

File: psi4/driver/qcdb/vib.py [hunk 4]
```diff
@@ -526,6 +541,24 @@ def vec_in_space(vec, space, tol=1.0e-4):
     xL = np.sqrt(reduced_mass_u) * wL
     vibinfo['x'] = QCAspect('normal mode', 'a0', xL, 'normalized un-mass-weighted')
 
+    # IR intensities, CCQC Proj. Eqns. 15-16
+    uconv_kmmol = (qcel.constants.get("Avogadro constant") * np.pi * 1.e-3 * qcel.constants.get("electron mass in u") *
+                   qcel.constants.get("fine-structure constant")**2 * qcel.constants.get("atomic unit of length") / 3)
+    uconv_D2A2u = (
+        qcel.constants.get('atomic unit of electric dipole mom.') * 1.e11 /
+        qcel.constants.get('hertz-inverse meter relationship') / qcel.constants.get('atomic unit of length'))**2
```
Comment (jturney) [inline psi4/driver/qcdb/vib.py]: Nice, I like the constants.

File: psi4/src/psi4/libmints/mintshelper.cc (hunk 3 above)
Comment (jturney) [reply psi4/src/psi4/libmints/mintshelper.cc]: Good call pulling the lambdas out of the loops.

**Pull request merged**